crates/wit-parser/src/resolve.rs

```diff
@@ -158,69 +191,48 @@ impl Resolve {
-        // Additionally note that the last item visited here is the root
-        // package, which is the one returned here.
-        let mut last = None;
-        let mut files = Vec::new();
-        let mut pkg = Some(pkg);
-        for name in order {
-            let pkg = deps.remove(&name).unwrap_or_else(|| pkg.take().unwrap());
-            files.extend(pkg.source_files().map(|p| p.to_path_buf()));
-            let pkgid = self.push(pkg)?;
-            last = Some(pkgid);
-        }
-
-        return Ok((last.unwrap(), files));
-
-        fn visit<'a>(
-            pkg: &'a UnresolvedPackage,
-            deps: &'a BTreeMap<PackageName, UnresolvedPackage>,
-            order: &mut IndexSet<PackageName>,
-            visiting: &mut HashSet<&'a PackageName>,
-        ) -> Result<()> {
-            if order.contains(&pkg.name) {
-                return Ok(());
-            }
-            pkg.source_map.rewrite_error(|| {
-                for (i, (dep, _)) in pkg.foreign_deps.iter().enumerate() {
-                    let span = pkg.foreign_dep_spans[i];
-                    if !visiting.insert(dep) {
-                        bail!(Error {
-                            span,
-                            msg: format!("package depends on itself"),
-                        });
-                    }
-                    if let Some(dep) = deps.get(dep) {
-                        visit(dep, deps, order, visiting)?;
-                    }
-                    assert!(visiting.remove(dep));
-                }
-                assert!(order.insert(pkg.name.clone()));
-                Ok(())
-            })
+        for pkg in pkgs {
+            let deps = path.join("deps");
+            let mut deps = self.parse_deps_dir(&deps).with_context(|| {
+                format!("failed to parse dependency directory: {}", deps.display())
+            })?;
+
+            // Perform a simple topological sort which will bail out on cycles
+            // and otherwise determine the order that packages must be added to
+            // this `Resolve`.
+            let mut order = IndexSet::new();
+            let mut visiting = HashSet::new();
+            for dep_pkg in deps.values().chain([&pkg]) {
+                visit(&dep_pkg, &deps, &mut order, &mut visiting)?;
+            }
+
+            // Using the topological ordering insert each package incrementally.
+            // Additionally note that the last item visited here is the root
+            // package, which is the one returned here.
+            let mut last = None;
+            let mut files = Vec::new();
+            let mut pkg = Some(pkg);
+            for name in order {
+                let pkg = deps.remove(&name).unwrap_or_else(|| pkg.take().unwrap());
+                files.extend(pkg.source_files().map(|p| p.to_path_buf()));
+                let pkgid = self.push(pkg)?;
+                last = Some(pkgid);
+            }
+
+            ret.push((last.unwrap(), files));
```

Comment:
This is me musing a bit here, but I wanted to highlight this a bit. No action needed in this PR, but figured I'd jot down my ideas.

Here the previous API returned the `PackageId` of the "top" package. That's being much more ambiguous after this PR because the concept of a "top" package is sort of gone where the top thing might have a whole bunch of packages. Once we buy into the concept of multiple "top" packages then some other related issues like https://github.com/bytecodealliance/wasm-tools/issues/1461 and https://github.com/bytecodealliance/wasm-tools/issues/1462 become more digestable (IMO at least).

AFAIK the only downside to having multiple "top" packages is the "world selection" process. Currently that allows the shorthand of specifying a single `foo` world name (or no name at all) and an appropriate world is selected. If we can reframe that functionality somehow to work with a set of "top" packages then it'd be pretty easy to implement those issues. That being said I don't know how best to rationalize how to select a world from a set of "top" packages.

I also bring this up because I can see here that maintaining a return value of the "top" packages is kinda awkward here and might be more awkward with one of my suggestions below of using `append` here. Anyway just wanted to muse a bit here in case you were inspired by something and had an idea.